invenio/ext/iiif/__init__.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015 CERN.
+#
+# Invenio is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Flask-IIIF extension."""
+
+from flask import request
+
+from flask_iiif import IIIF
+from flask_iiif.errors import MultimediaError
+
+from invenio.modules.documents.utils import (
+    identifier_to_path, identifier_to_path_and_permissions
+)
+from werkzeug.exceptions import (
+    Forbidden, NotFound
+)
+
+__all__ = ('setup_app', )
+
+iiif = IIIF()
+
+
+def setup_app(app):
+    """Setup Flask-IIIF extension."""
+    try:
+        assert (
+            'invenio.modules.documents' not in
+            app.config.get('PACKAGES_EXCLUDE')
+        )
+    except AssertionError:
+        raise MultimediaError(
+            "Could not initialize the Flask-IIIF extension because "
+            ":class:`~invenio.modules.documents.api:Document` is missing"
+        )
+    else:
+        iiif.init_app(app)
+        iiif.init_restful(app.extensions['restful'])
+        iiif.uuid_to_path_handler(identifier_to_path)
+        app.config['IIIF_CACHE_HANDLER'] = 'invenio.ext.cache:cache'
+
+        # Make sure that permissions are applied on cached images as well.
+
+        @app.before_request
+        def check_cached_permissions():
+            if 'iiifimageapi' == request.endpoint:
+                uuid = request.view_args.get('uuid')
+                try:
+                    identifier_to_path_and_permissions(uuid)
+                except (Forbidden, NotFound):
+                    raise
+                else:
```

Comment:
if the image is cached then it's not passing through the `uuid_to_path` callback it takes it directly from the cache and exposes it